File: 2_avoha/backend/src/routes/sse.ts
import type { FastifyInstance } from "fastify";
import { requireSession } from "../lib/auth-guard.js";
import { publish, subscribe, type InventoryEvent } from "../lib/sse-bus.js";

const HEARTBEAT_MS = 25_000;

function write(reply: import("fastify").FastifyReply, ev: InventoryEvent): void {
  reply.raw.write(`data: ${JSON.stringify(ev)}\n\n`);
}

export async function sseRoutes(app: FastifyInstance) {
  app.get("/sse/inventory", async (req, reply) => {
    const userId = await requireSession(req, reply);
    if (!userId) return;

    reply.raw.writeHead(200, {
      "Content-Type": "text/event-stream",
      "Cache-Control": "no-cache, no-transform",
      Connection: "keep-alive",
      "X-Accel-Buffering": "no",
    });
    reply.raw.write("retry: 5000\n\n");
    reply.hijack();

    const unsubscribe = subscribe(userId, (ev) => write(reply, ev));
    const heartbeat = setInterval(() => {
      try {
        reply.raw.write(`: ping\n\n`);
      } catch {
        /* already closed */
      }
    }, HEARTBEAT_MS);

    const cleanup = () => {
      clearInterval(heartbeat);
      unsubscribe();
    };

    req.raw.on("close", cleanup);
    req.raw.on("error", cleanup);

    // hello event (클라이언트가 연결 확인용으로 활용)
    publish(userId, { type: "ping" });
  });
}

File: 2_avoha/backend/app/routes/sse.py
from __future__ import annotations

import asyncio
import json
import uuid
from typing import AsyncIterator

from fastapi import APIRouter, Depends, Request
from sse_starlette.sse import EventSourceResponse

from app.deps import require_user
from app.services import sse_bus

HEARTBEAT_S = 25

router = APIRouter()


@router.get("/sse/inventory")
async def inventory_stream(
    request: Request,
    user_id: uuid.UUID = Depends(require_user),
) -> EventSourceResponse:
    queue = sse_bus.subscribe(user_id)

    async def event_gen() -> AsyncIterator[dict[str, str]]:
        try:
            # hello event — 연결 직후 확인용
            yield {"data": json.dumps({"type": "ping"})}

            while True:
                if await request.is_disconnected():
                    break
                try:
                    ev = await asyncio.wait_for(queue.get(), timeout=HEARTBEAT_S)
                except asyncio.TimeoutError:
                    # sse-starlette 가 콜론 주석으로 heartbeat 보내도록 ping_message 를 yield
                    yield {"event": "ping", "data": ""}
                    continue
                yield {"data": json.dumps(ev)}
        finally:
            sse_bus.unsubscribe(user_id, queue)

    return EventSourceResponse(event_gen(), ping=HEARTBEAT_S)
